cdc/log.cc

```diff
@@ -416,6 +416,194 @@ utils::UUID generate_timeuuid(api::timestamp_type t) {
     return utils::UUID_gen::get_random_time_UUID_from_micros(t);
 }
 
+struct atomic_column_update {
+    column_id id;
+    atomic_cell cell;
+};
+
+struct nonatomic_column_deletion {
+    column_id id;
+};
+
+struct nonatomic_column_update {
+    column_id id;
+    std::vector<std::pair<bytes, atomic_cell>> cells;
+};
+
+struct static_row_update {
+    gc_clock::rep ttl;
+    std::vector<atomic_column_update> atomic_entries;
+    std::vector<nonatomic_column_deletion> nonatomic_deletions;
+    std::vector<nonatomic_column_update> nonatomic_updates;
+};
+
+struct clustered_row_insert {
+    gc_clock::rep ttl;
+    clustering_key key;
+    std::vector<atomic_column_update> atomic_entries;
+    std::vector<nonatomic_column_deletion> nonatomic_deletions;
+};
+
+struct clustered_row_update {
+    gc_clock::rep ttl;
+    clustering_key key;
+    std::vector<atomic_column_update> atomic_entries;
+    std::vector<nonatomic_column_deletion> nonatomic_deletions;
+    std::vector<nonatomic_column_update> nonatomic_updates;
```

Comment:
How come insert doesn't have a `nonatomic_inserts` or equivalent?

Reply:
`nonatomic_updates` are for adding cells to nonatomic columns. E.g. suppose you have a `map<int, int> map_col`. You can perform:
```
update ... set map_col = map_col + {1:2}
```
This cannot be expressed in an insert.

In insert you can replace the entire value
```
insert into ... (..., map_col) values (..., {1:2, 3:4})
```
which underneath is translated to a tombstone with timestamp `X-1` and cells with timestamp `X`. So I split the above to two changes with different timestamps: one will be an insert with `nonatomic_deletion`, the other will be an update with `nonatomic_updates`.